Add handling for unmarshalling unions with enums.

ostromart opened the pull request:
Unions with enums depend on new type resolution in the generated code. 
Enum types are strings in JSON source tree. To determine whether they can be represented as enums, the list of all possible enum types and their values is needed from the local schema context. 
If any of these values match, the field is cast as the corresponding type that implements the union interface. Otherwise, unmarshal is attempted into the other, non-enum types.

robshakir commented on ytypes/leaf.go:
```diff
@@ -508,117 +509,203 @@ the resulting Bgp_Neighbor_RouteReflector would have field
 		return fmt.Errorf("%s is not a valid field name in %T", fieldName, parent)
 	}
-	ft, err := getFieldElemType(parent, fieldName)
+	dft, _ := parentT.Elem().FieldByName(fieldName)
```
What is the `_` here? AFAICS `FieldByName` only returns one `reflect.Value`?

ostromart replied:
This is the method for the Type (which doesn't come up in godoc, seems like a rendering bug):
// FieldByName returns the struct field with the given name
        // and a boolean indicating if the field was found.
        FieldByName(name string) (StructField, bool)
We don't need to check ok because we known it exists from the line above.

robshakir replied:
Ah, thanks -- strange, I looked a GoDoc, weird that it has a bug here.

robshakir commented on ytypes/leaf.go:
```diff
@@ -508,117 +509,203 @@ the resulting Bgp_Neighbor_RouteReflector would have field
-		return reflect.TypeOf(nil), fmt.Errorf("%s is not a valid field name in %T", fieldName, parent)
+// schemaToEnumTypes returns the actual enum types (rather than the interface
+// type) for a given schema, which must be for an enum type.
+func schemaToEnumTypes(schema *yang.Entry, t reflect.Type) ([]reflect.Type, error) {
```
Assume here that `t` must be the struct that contains the enum field? Consider describing this in the comment above.

ostromart replied:
Done.

robshakir commented on ytypes/leaf.go:
```diff
@@ -508,117 +509,203 @@ the resulting Bgp_Neighbor_RouteReflector would have field
+	util.DbgPrint("path is %s for schema %s", ygen.EntrySchemaPath(schema), schema.Name)
 
-	return reflect.TypeOf(nil), fmt.Errorf("%T is not a valid parent type", parent)
+	return enumTypesMap[ygen.EntrySchemaPath(schema)], nil
```
What happens here if `enumTypeMap` doesn't contain an entry with the relevant path? Should this be a non-error case for the function -- I think not. Consider checking whether the key exists and returning an error if it doesn't.

ostromart replied:
I had originally intended to handle this separately by returning nil but apparently didn't in the end. Good catch.

ostromart replied:
It's not an error. The function that calls this aggregates all possible union values, which can include non-enum types. If this returns nil, it just means that the union doesn't contain enum values.

robshakir replied:
Ah, understood.

robshakir commented on ytypes/leaf_test.go:
```diff
@@ -865,6 +895,7 @@ func TestUnmarshalLeaf(t *testing.T) {
 			want: LeafContainerStruct{Uint64Leaf: ygot.Uint64(42)},
 		},
 		{
+			// TODO DEBUG 8
```
Remove comment?

ostromart replied:
Done.

robshakir commented on ytypes/leaf_test.go:
```diff
@@ -880,6 +911,17 @@ func TestUnmarshalLeaf(t *testing.T) {
 			want: LeafContainerStruct{UnionLeaf: &UnionLeafType_Uint32{Uint32: 42}},
 		},
 		{
+			// TODO DEBUG 11
```
Remove comment?

ostromart replied:
Done.

ostromart commented:
Note that I removed the ygen dependency and moved the schema to path translation to a local function. 
Couple of questions around that:
 - why is choice/case not part of the the path?
 - traverseElementSchemaPath doesn't seem to consider fakeroot i.e. I don't see the fakeroot in the EnumTypes map but there's no check for it in that function. I thought there's a fakeroot element in the schema now? At least I am testing for it in isFakeRoot.
 - maybe consider streamlining the function? The reversing seems unnecessary and the fact that you are converting to absolute path may be more meaningfully expressed by prepending "/" to the resulting string.

robshakir commented:
1) `choice` and `case` statements generally aren't included in the schema path since there must be uniqueness in the top-level names of the `case` statements of a choice. So this function skips them.
2) In the parsed `yang.Entry` that `goyang/pkg/yang` creates, there is no fake root. This is added by `ygen`. In the `yang.Entry` tree that is serialised for use in `ytypes`, the schema tree *is* included such that that element exists for parsing of the fake root. The tree in `goyang/pkg/yang` has multiple roots (one per module). To get between modules, you need to use the module set that the parsed module was associated with.
3) Not entirely sure how to avoid the reversing. This function starts at the leaf and traverses back to the root -- since we don't know how many parents there are, it seems like pushing these names into a slice and then reversing it is a reasonable strategy? 
3) For the absolute path, you're referencing `goyangPathToSchemaPath` right? `traverseElementSchemaPath` walks up the tree and does provide the module name in the path, so isn't quite an absolute path - `slicePathToString` is not required to be absolute. Only `goyangPathToSchemaPath` is absolute, and is prefixed by `/` by adding `""` to the start of the string slice.
